cranlock/lock.py

```diff
@@ -0,0 +1,226 @@
+import argparse
+from bs4 import BeautifulSoup
+from enum import Enum
+from functools import reduce
+import os
+import re
+import requests
+
+CRAN_PACKAGE_URL_FORMAT = "https://cran.r-project.org/web/packages/{package}/index.html"
+
+
+def get_cran_url(package_name: str) -> str:
+    return CRAN_PACKAGE_URL_FORMAT.format(package=package_name)
+
+
+def get_table_row(table, td_text):
+    """ Find a row in a table that has td_text as one column's text """
+    td = table.find('td', string=td_text)
+
+    if td is None:
+        return None
+
+    return td.find_parent('tr')
+
+
+# Cache the responses for all CRAN URLs we have visited, since a package often appears more than
+# once in a dependency tree.
+url_cache = dict()
+
+
+def get_info_table(package: str):
+    """Get the HTML table from the CRAN site for a package, which holds all the dependency info"""
+    url = get_cran_url(package)
+    if url_cache.get(url, None):
+        response = url_cache[url]
+    else:
+        response = requests.get(url)
+        url_cache[url] = response
+
+    if response.status_code != 200:
+        raise Exception("Package {} does not exist on CRAN".format(package))
+
+    soup = BeautifulSoup(response.text, 'html.parser')
+
+    # Dependency information lives in an html table called 'Package $PKGNAME summary'
+    return soup.find('table', summary=re.compile('Package [\\w\\W]+ summary'))
+
+
+def get_dependencies(package: str):
+    """ Get all of the packages that a package `depends` or `imports` whose names are links.
+    (If their names aren't links, they're base packages that we don't need to worry about)
+    """
+
+    table = get_info_table(package)
+
+    imports_row = get_table_row(table, "Imports:")
+    depends_row = get_table_row(table, "Depends:")
+
+    imports = imports_row.find_all('a') if imports_row is not None else []
+    depends = depends_row.find_all('a') if depends_row is not None else []
+
+    return list(map(lambda a: a.string, imports + depends))
+
+
+# Cache the dependencies for every dependency, because packages often appear multiple times in
+# the depedency tree.
+dependencies_cache = dict()
+
+
+def get_all_dependencies(package: str):
+    """Recursively finds all the dependencies of a package, and builds up a dependency tree"""
+    if dependencies_cache.get(package, None) is not None:
+        return dependencies_cache[package]
+
+    first_level_dependencies = get_dependencies(package)
+
+    if first_level_dependencies == []:
+        return []
```

Comment:
I don't think you need this block because `reduce` will use the initalizer as a default if the iterable is empty. Also this block means that sometimes this function returns a list but other times a dict.